design-proposals/network-device-capacity.md

```diff
@@ -0,0 +1,148 @@
+It will be available as a domain-resources-requests entry.
+
+#### Validation
+- Should be immutable.
+- Should accept an integer value between 0 and 32.
+- Should be accepted only as `requests` (not as `limits`).
+- In case the value is less than the amount of interfaces defined already,
+  it has no effect.
+
+#### VMI Spec example
+```yaml
+
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-iface-request-example
+spec:
+  domain:
+    devices:
+      disks:
+        - disk:
+            bus: virtio
+          name: containerdisk
+      interfaces:
+        - masquerade: {}
+          name: default
+        - bridge: {}
+          name: secondary
+    resources:
+      requests:
+        interface: 5
+      rng: {}
+  networks:
+    - name: default
+      pod: {}
+    - multus:
+        networkName: some-nad
+      name: secondary 
+  volumes:
+    - containerDisk:
+        image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+      name: containerdisk
+```
+
+## Design Details
+The solution is taking advantage of the existing libvirt capability to auto-generate
+the needed configuration (e.g. controllers) when specifying network interfaces.
+
+The main change is in the logic of creating a domain in the virt-launcher.
+
+This is the main flow suggested:
+- Base on the resources-requests-interface and the current defined interfaces, generate
+  [interface-placeholders](#network-interface-placeholder) and add them to the domain spec.
+- Following the creation of the domain configuration, remove these placeholders
+  from he configuration and start the domain without them.
+- At this stage, the domain has been started without the additional interface devices
+  but with the additional configuration (e.g. controllers) they originally generated.
+
+#### Network Interface Placeholder
+The interface placeholder is a temporary dummy definition of an interface in the domain spec
+of libvirt. It is used to trigger libvirt to auto-generate dependent devices (e.g. PCIe controllers).
+
+After the domain configuration is constructed, these placeholders are removed.
+
+This is an example of such a placeholder:
+```xml
+<interface type='ethernet'>
+  <target dev='placeholder-1' managed='no'/>
+  <model type='virtio-non-transitional'/>
+  <mtu size='1500'/>
+  <rom enabled='no'/>
+</interface>
+```
+
+## Implementation Phases
+N/A
+
+## Appendix A: Alternatives
+There have been several other implementation alternatives to answer the goals.
+This section lists the alternatives to allow an open discussion and record other proposals.
+
+- Allow users to define explicitly domain controllers.
+
+  This option includes the full exposure of the controllers devices details or a partial
+  abstraction of it.
+
+- While using the same suggested API towards the user, construct the required controllers
+  directly (and not through network interface placeholders).
+
+- Use a PCI bridge.
```

Comment:
@andreabolognani can you please help me with this one?
I understood from @AlonaKaplan that some D/S projects are using this option.

Reply:
A PCI bridge (specifically `pcie-to-pci-bridge` device) would provide a suitable number of hotplug-capable slots, but they would be traditional PCI slots instead of PCI Express slots.

In general, while libvirt will allow manually plugging a virtio device into a traditional PCI slot on a PCI Express based machine such as q35, it will never decide on such a placement on its own. So, even if the bridge was present in the configuration, libvirt would ignore it and keep adding / expecting `pcie-root-port` controllers instead.

I would be interested in more information about the projects using PCI bridges to implement hotplug. @AlonaKaplan, can you please provide some pointers?

Reply:
It is an u/s project - Genesis Cloud.
The contact I have is Philipp Riederer priederer@genesiscloud.com.

Reply:
They appear to be @toelke on GitHub.

I've looked at [Genesis Cloud's KubeVirt fork](https://github.com/genesiscloud/kubevirt) and I've been unable to find any mention of PCI bridges anywhere.

Reply:
What we do is a little more funky and I am still in internal discussion how we can share what we have. I am allowed to talk, but not to show code right now.

We are using PCI "slots" for storage hotplug from a `vhost-user` socket. These allow hot-plugging controllers without pre-allocating them.

We run a sidecar hook that ensures that this config exists in the domain XML on startup:

```xml
<devices>
  <controller type="pci" index=0, model="pcie-root"/>
  <controller type="pci" index=1, model="pcie-root-port"/>
  <controller type="pci" index=2, model="pcie-to-pci-bridge"/>
  <controller type="pci" index=3, model="pci-bridge"/>
</devices>
```

This sidecar then listens to RPC calls that we send when we want to do a hotplug (we bypass the kubevirt hotplug there). It uses the libvirt API (`domain.attachDeviceFlags`) to attach such a snippet to the VM:

```xml
  <disk type="vhostuser" device="disk" model="virtio-non-transitional" snapshot="no">
    <driver name="qemu" type="raw"/>
    <source type="unix" path="/path/to/vhost-user.sock"/>
    <target dev="vda" bus="virtio"/>
    <address type="pci" domain="0" bus="3" slot="1" function="0"/>
  </disk>
```

(We would obviously be counting up the `disk.address.slot` with each subsequent hotplug).

I am sure this could be also be done for network devices.

Reply:
You can also query the xml from libvirt and collect all `<disk/>`s that are currently attached to "your" PCI controller and find an unused slot that way; more work for the sidecar but less bookkeeping.

Reply:
Right, but that still requires managing PCI addresses manually, which AIUI is something that KubeVirt doesn't want to be in the business of doing.

There's also the (possibly minor) concern that, with this approach, the hotplugged devices will look different from the initial ones inside the guest OS: the latter will be PCI Express devices, while the former will be conventional PCI devices, on account of having been plugged into different kinds of slots. This will result in things such as different PCI vendor:device IDs.

Reply:
In our setup all storage is plugged in using this scheme -- which is a special case.

Reply:
I have not added any further text here and for now removed the TBD note.
Perhaps it will be good to add some basic explanation to this option and if you can help me phrase it in a short sentence/paragraph it will be great.